### PR TITLE
ATLAS-4111: Add replicationFactor attribute to kafka topic entity

### PR DESCRIPTION
**Pull request by bmaidics**
Currently, the replicationFactor is missing from kafka topic entities in Atlas. We should add them.

### CODE REVIEW

**Review by urbandan — APPROVED** [review]
LGTM

File: addons/models/1000-Hadoop/patches/018-kafka_topic_add_rf_attribute.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      "id": "TYPEDEF_PATCH_1000_018_001",
+      "description": "Add 'replicationFactor' attribute to kafka_topic",
+      "action": "ADD_ATTRIBUTE",
+      "typeName": "kafka_topic",
+      "applyToVersion": "1.7",
+      "updateToVersion": "1.8",
+      "params": null,
+      "attributeDefs": [
+        {
+          "name": "replicationFactor",
```
Comment (sarathsubramanian) [inline addons/models/1000-Hadoop/patches/018-kafka_topic_add_rf_attribute.json:13]: can you also add this new attribute to the base 'kafka_topic' model - 1070-kafka_model.json?

File: common/src/main/java/org/apache/atlas/utils/KafkaUtils.java [hunk 2]
```diff
@@ -126,15 +126,7 @@ public Integer getPartitionCount(String topicName) throws ExecutionException, In
...
-        Integer partitionCount = null;
-        DescribeTopicsResult describeTopicsResult =  adminClient.describeTopics(Collections.singleton(topicName));
-        Map<String, KafkaFuture<TopicDescription>> futureMap = describeTopicsResult.values();
-        for(Map.Entry<String, KafkaFuture<TopicDescription>> futureEntry : futureMap.entrySet()) {
-            KafkaFuture<TopicDescription> topicDescriptionFuture = futureEntry.getValue();
-            TopicDescription topicDescription = topicDescriptionFuture.get();
-            List<TopicPartitionInfo> partitionList = topicDescription.partitions();
-            partitionCount = partitionList.size();
-        }
+        Integer partitionCount = getPartitionList(topicName).size();
```
Comment (sarathsubramanian) [inline common/src/main/java/org/apache/atlas/utils/KafkaUtils.java]: getPartitionList(topicName) can possibly return null. consider adding null check

File: common/src/main/java/org/apache/atlas/utils/KafkaUtils.java [hunk 3]
```diff
@@ -143,6 +135,33 @@ public Integer getPartitionCount(String topicName) throws ExecutionException, In
         return partitionCount;
     }
 
+    public Integer getReplicationFactor(String topicName) throws ExecutionException, InterruptedException {
+        if(LOG.isDebugEnabled()) {
+            LOG.debug("==> KafkaUtils.getReplicationFactor({})", topicName);
+        }
+
+        Integer replicationFactor = getPartitionList(topicName).stream().mapToInt(x -> x.replicas().size()).max().getAsInt();
```
Comment (sarathsubramanian) [inline common/src/main/java/org/apache/atlas/utils/KafkaUtils.java]: consider adding null check for getPartitionList(topicName)

File: common/src/main/java/org/apache/atlas/utils/KafkaUtils.java [hunk 4]
```diff
@@ -143,6 +135,33 @@ public Integer getPartitionCount(String topicName) throws ExecutionException, In
...
+        }
+
+        return replicationFactor;
+    }
+
+    private List<TopicPartitionInfo> getPartitionList(String topicName) throws ExecutionException, InterruptedException {
+        List<TopicPartitionInfo> partitionList = null;
+
+        DescribeTopicsResult describeTopicsResult =  adminClient.describeTopics(Collections.singleton(topicName));
+        Map<String, KafkaFuture<TopicDescription>> futureMap = describeTopicsResult.values();
```
Comment (sarathsubramanian) [inline common/src/main/java/org/apache/atlas/utils/KafkaUtils.java]: can 'describeTopicsResult' be null if no topics exist? add necessary null check

**Comment by bmaidics** [comment]
Thanks @sarathsubramanian for the review, I updated the PR with the changes.

**Comment by sarathsubramanian** [comment]
changes looks good. I merged the PR.

**Comment by sarathsubramanian** [comment]
please close the PR.

**Pull request closed without merging**